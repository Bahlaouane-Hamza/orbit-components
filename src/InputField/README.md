# InputField
To implement InputField component into your project you'll need to add the import:
```jsx
import InputField from "@kiwicom/orbit-components/lib/InputField";
```
After adding import into your project you can use it simply like:
```jsx
<InputField />
```
## Props
Table below contains all types of the props available in InputField component.

| Name          | Type              | Default      | Description                      |
| :------------ | :---------------- | :----------- | :------------------------------- |
| disabled      | `boolean`         |              | If `true`, the InputField will be disabled.
| dataTest      | `string`          |              | Optional prop for testing purposes.
| error         | `React.Node`      |              | The error to display beneath the InputField. [See Functional specs](#functional-specs)
| help          | `React.Node`      |              | The help to display beneath the InputField.
| label         | `string`          |              | The label for the InputField. [See Functional specs](#functional-specs)
| inlineLabel   | `boolean`         |              | If true the label renders on the left side of input
| maxLength     | `number`          |              | Specifies the maximum number of characters allowed.
| maxValue      | `number`          |              | Specifies the maximum value for the InputField.
| minLength     | `number`          |              | Specifies the minimum number of characters allowed.
| minValue      | `number`          |              | Specifies the minimum value for the InputField.
| required      | `boolean`         |              | If true, the label is displayed as required.
| name          | `string`          |              | The name for the InputField.
| onBlur        | `func`            |              | Function for handling onBlur event.
| onChange      | `func`            |              | Function for handling onClick event.
| onFocus       | `func`            |              | Function for handling onFocus event.
| onKeyDown     | `func`            |              | Function for handling onKeyDown event.
| onKeyUp       | `func`            |              | Function for handling onKeyUp event.
| placeholder   | `string`          |              | The placeholder of the InputField.
| **prefix**    | `React.Node`      |              | The prefix component for the InputField. 
| ref           | `func`            |              | Prop for forwarded ref of the InputField. [See Functional specs](#functional-specs)
| **size**      | [`enum`](#enum)   | `"normal"`   | The size of the InputField.
| suffix        | `React.Node`      |              | The suffix component for the InputField.
| **type**      | [`enum`](#enum)   | `"text"`     | The type of the InputField.
| value         | `string`          |              | Specifies the value of the InputField.

### enum

| type         | size        |
| :----------- | :---------- |
| `"text"`     | `"small"`   |
| `"number"`   | `"normal"`  |
| `"email"`    |
| `"password"` |


## Functional specs
* The `error` prop overwrites the `help` prop, due to higher priority.

* The color of the label will turn into cloud shade when the InputField has some filled value.

* You can use `string` for currency InputField, or `React.Node` for InputField with icon.

* `ref` can be used for example auto-focus the elements immediately after render.
```jsx
class Component extends React.PureComponent<Props> {

  constructor() {
    super()
    this.ref = React.createRef();
  }
  
  componentDidMount() {
    this.ref.current && this.ref.current.focus();
  }

  ref: { current: React$ElementRef<*> | null };

  render() {
    return (
      <InputField ref={this.ref} />
    )
  }
}
```
