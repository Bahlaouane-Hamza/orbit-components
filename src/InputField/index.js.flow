// @flow
import type { Globals, Ref } from "../common/common.js.flow";

export type Size = "small" | "normal";

export type Props = {|
  +size?: Size,
  +type?: "text" | "number" | "email" | "password",
  +name?: string,
  +label?: string,
  +inlineLabel?: boolean,
  +value?: string | number,
  +placeholder?: string,
  +prefix?: React$Node,
  +suffix?: React$Node,
  +help?: React$Node,
  +error?: React$Node,
  +disabled?: boolean,
  +maxValue?: number,
  +minValue?: number,
  +maxLength?: number,
  +minLength?: number,
  required?: boolean,
  +onChange?: (ev: SyntheticInputEvent<HTMLInputElement>) => void | Promise<any>,
  +onFocus?: (ev: SyntheticInputEvent<HTMLInputElement>) => void | Promise<any>,
  +onBlur?: (ev: SyntheticInputEvent<HTMLInputElement>) => void | Promise<any>,
  +onKeyDown?: (ev: SyntheticKeyboardEvent<HTMLInputElement>) => void | Promise<any>,
  +onKeyUp?: (ev: SyntheticKeyboardEvent<HTMLInputElement>) => void | Promise<any>,
  ...Globals,
  ...Ref,
|};

declare export var FakeInput: React$ComponentType<>;

declare export var Input: React$ComponentType<>;

declare export var Prefix: React$ComponentType<>;

declare export var InputContainer: React$ComponentType<>;

declare export default React$ComponentType<Props>;
