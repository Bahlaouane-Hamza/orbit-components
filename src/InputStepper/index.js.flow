// @flow
import type { Globals } from "../common/common.js.flow";
import type { Size } from "../InputField/index";

export type Props = {|
  ...Globals,
  +size?: Size,
  +label?: string,
  +defaultValue?: number,
  +step?: number,
  +help?: React$Node,
  +error?: React$Node,
  +name?: string,
  +disabled?: boolean,
  +maxValue?: number,
  +minValue?: number,
  +required?: boolean,
  +onChange?: number => void | Promise<any>,
  +onFocus?: (ev: SyntheticInputEvent<HTMLInputElement>) => void | Promise<any>,
  +onBlur?: (ev: SyntheticInputEvent<HTMLInputElement>) => void | Promise<any>,
|};

export type State = {|
  value: number,
|};

declare export default React$ComponentType<Props>;
