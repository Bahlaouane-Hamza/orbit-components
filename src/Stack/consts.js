// @flow

export const DIRECTIONS = {
  ROW: "row",
  COLUMN: "column",
};

export const ALIGNS = {
  START: "start",
  END: "end",
  CENTER: "center",
  EVEN: "even",
};

export const SPACINGS = {
  EXTRATIGHT: "extraTight",
  TIGHT: "tight",
  CONDENSED: "condensed",
  COMPACT: "compact",
  NATURAL: "natural",
  COMFY: "comfy",
  LOOSE: "loose",
  EXTRALOOSE: "extraLoose",
};

export const TOKENS = {
  DISTRIBUTION: "distribution",
  ALIGN: "align",
  SPACING: "spacing",
  MOBILESPACING: "mobileSpacing",
};
