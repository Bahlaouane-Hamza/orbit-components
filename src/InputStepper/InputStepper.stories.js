// @flow
import * as React from "react";
import { storiesOf, setAddon } from "@storybook/react";
import styles from "@sambego/storybook-styles";
import { action } from "@storybook/addon-actions";
import chaptersAddon from "react-storybook-addon-chapters";
import { withKnobs, text, number, select, boolean } from "@storybook/addon-knobs/react";

import { SIZE_OPTIONS } from "../InputField/consts";

import InputStepper from "./index";

setAddon(chaptersAddon);

storiesOf("InputStepper", module)
  .addDecorator(withKnobs)
  .addDecorator(
    styles({
      padding: "20px",
    }),
  )

  .addWithChapters("Default", () => ({
    info: "Some description about this type of InputStepper in general.",
    chapters: [
      {
        sections: [
          {
            sectionFn: () => <InputStepper onChange={action("onChange")} />,
          },
        ],
      },
    ],
  }))

  .addWithChapters("With help", () => {
    const label = text("Label", "Aduls");
    const help = text("help", "You need to enter count of adults");
    const size = select("Size", Object.values(SIZE_OPTIONS), SIZE_OPTIONS.NORMAL);

    return {
      info: "Here you can try InputStepper component with additional functionality.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <InputStepper
                  label={label}
                  size={size}
                  help={help}
                  onChange={action("onChange")}
                  onFocus={action("onFocus")}
                  onBlur={action("onBlur")}
                />
              ),
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("With different size", () => {
    const label = text("Label", "Label");
    const size = select("Size", Object.values(SIZE_OPTIONS), SIZE_OPTIONS.NORMAL);

    return {
      info: "Here you can try InputStepper component with additional functionality.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <InputStepper
                  label={label}
                  size={size}
                  onChange={action("onChange")}
                  onFocus={action("onFocus")}
                  onBlur={action("onBlur")}
                />
              ),
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Playground", () => {
    const min = number("minValue", 1);
    const max = number("maxValue", 10);
    const step = number("step", 2);
    const defaultValue = number("defaultValue", 4);
    const label = text("Label", "Label");
    const size = select("Size", Object.values(SIZE_OPTIONS), SIZE_OPTIONS.NORMAL);
    const help = text("Help", undefined);
    const name = text("Name", "name");
    const error = text("Error", "Error message (explain how to solve it)");
    const disabled = boolean("disabled", false);
    const required = boolean("required", false);
    const dataTest = text("dataTest", "test");
    return {
      info: "Some description about this type of InputStepper in general.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <InputStepper
                  label={label}
                  defaultValue={defaultValue}
                  size={size}
                  step={step}
                  error={error}
                  help={help}
                  name={name}
                  maxValue={max}
                  minValue={min}
                  disabled={disabled}
                  required={required}
                  dataTest={dataTest}
                  onChange={action("onChange")}
                  onFocus={action("onFocus")}
                  onBlur={action("onBlur")}
                />
              ),
            },
          ],
        },
      ],
    };
  });
