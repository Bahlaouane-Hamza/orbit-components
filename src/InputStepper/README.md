# InputStepper

To implement InputStepper component into your project you'll need to add the import:

```jsx
import InputStepper from "@kiwicom/orbit-components/lib/InputStepper";
```

After adding import into your project you can use it simply like:

```jsx
<InputStepper />
```

## Props

Table below contains all types of the props available in InputStepper component.

| Name            | Type             | Default     | Description                                        |
| :-------------- | :--------------- | :---------- | :------------------------------------------------- |
| defaultValue    | `number`         | `0`         | Specifies the value of the InputStepper. [See Functional specs](#functional-specs)
| disabled        | `boolean`        | `false`     | If `true`, the InputStepper will be disabled.
| error           | `React.Node`     |             | The error to display beneath the InputStepper. [See Functional specs](#functional-specs)
| help            | `React.Node`     |             | The help to display beneath the InputStepper.
| label           | `string`         |             | The label for the InputStepper. [See Functional specs](#functional-specs)
| maxValue        | `number`         | `∞`         | Specifies the maximum value for the InputStepper.
| minValue        | `number`         | `-∞`        | Specifies the minimum value for the InputStepper.
| name            | `string`         |             | The name for the InputStepper.
| required        | `boolean`        | `false`     | If `true`, the label is displayed as required.
| onBlur          | `func`           |             | Function for handling onBlur event.
| onChange        | `func`           |             | Function for handling onClick event.
| onFocus         | `func`           |             | Function for handling onFocus event.
| ref             | `func`           |             | Prop for forwarded ref of the InputStepper. [See Functional specs](#functional-specs)
| size            | [`enum`](#enum)  | `"normal"`  | The size of the InputStepper.
| step            | `number`         | `1`         | Specifies the value of step to increment and decrement.

### enum

| size       |
| :--------- |
| `"small"`  |
| `"normal"` |

## Functional specs

* The `error` prop overwrites the `help` prop, due to higher priority.

* The color of the label will turn into cloud shade when the InputStepper has some filled value.

* The prop `defaultValue` sets up the default value when component mounts. If you need to get the current and validated value of InputStepper, use arrow function for it,  e.g.:
```jsx
<InputStepper onChange={(value) => doSomething(value)} />
```

* `ref` can be used for example auto-focus the elements immediately after render.
```jsx
class Component extends React.PureComponent<Props> {

  constructor() {
    super()
    this.ref = React.createRef();
  }
  
  componentDidMount() {
    this.ref.current && this.ref.current.focus();
  }

  ref: { current: React$ElementRef<*> | null };

  render() {
    return (
      <InputStepper ref={this.ref} />
    )
  }
}
```
