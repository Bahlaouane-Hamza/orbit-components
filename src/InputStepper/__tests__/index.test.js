// @flow
import * as React from "react";
import { shallow } from "enzyme";

import InputStepper from "../index";
import { SIZE_OPTIONS } from "../../InputField/consts";

describe(`InputField with help, prefix and suffix`, () => {
  const size = SIZE_OPTIONS.NORMAL;
  const label = "Label";
  const defaultValue = 1;
  const step = 2;
  const help = "Help message";
  const error = "Error message";
  const name = "name";
  const disabled = false;
  const maxValue = 100;
  const minValue = 1;
  const required = false;
  const onChange = jest.fn();
  const onFocus = jest.fn();
  const onBlur = jest.fn();

  const component = shallow(
    <InputStepper
      size={size}
      label={label}
      defaultValue={defaultValue}
      step={step}
      help={help}
      error={error}
      name={name}
      disabled={disabled}
      maxValue={maxValue}
      minValue={minValue}
      required={required}
      onChange={onChange}
      onFocus={onFocus}
      onBlur={onBlur}
    />,
  );
  const input = component.find("InputField");
  const InitialState = { value: 0 };

  it("Initialize with the state", () => {
    component.setState(InitialState);
    expect(component.state()).toEqual(InitialState);
  });

  it("should contain an input", () => {
    expect(input.exists()).toBe(true);
  });

  it("should have passed props", () => {
    expect(input.prop("size")).toBe(size);
    expect(input.prop("label")).toBe(label);
    expect(input.prop("value")).toBe(defaultValue);
    expect(input.prop("help")).toBe(help);
    expect(input.prop("error")).toBe(error);
    expect(input.prop("disabled")).toBe(disabled);
    expect(input.prop("maxValue")).toBe(maxValue);
    expect(input.prop("minValue")).toBe(minValue);
    expect(input.prop("required")).toBe(required);
  });

  it("should execute onChange method", () => {
    input.simulate("change");
    expect(onChange).toHaveBeenCalled();
  });

  it("should execute onFocus method", () => {
    input.simulate("focus");
    expect(onFocus).toHaveBeenCalled();
  });

  it("should execute onBlur method", () => {
    input.simulate("focus");
    input.simulate("blur");
    expect(onBlur).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
