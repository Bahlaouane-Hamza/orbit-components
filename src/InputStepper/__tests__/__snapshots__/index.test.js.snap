// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField with help, prefix and suffix should match snapshot 1`] = `
<InputStepper__StyledInputStepper>
  <InputField
    disabled={false}
    error="Error message"
    help="Help message"
    label="Label"
    maxValue={100}
    minValue={1}
    name="name"
    onBlur={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onChange={[Function]}
    onFocus={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onKeyDown={[Function]}
    prefix={
      <ButtonLink
        component={[Function]}
        disabled={true}
        external={false}
        iconLeft={
          <MinusCircle
            color="secondary"
          />
        }
        onClick={[Function]}
        size="normal"
        transparent={true}
        type="primary"
        width={0}
      />
    }
    required={false}
    size="normal"
    suffix={
      <ButtonLink
        component={[Function]}
        disabled={false}
        external={false}
        iconLeft={
          <PlusCircle
            color="secondary"
          />
        }
        onClick={[Function]}
        size="normal"
        transparent={true}
        type="primary"
        width={0}
      />
    }
    type="number"
    value={0}
  />
</InputStepper__StyledInputStepper>
`;
